### PR TITLE
[Issue-24-Partial] Archive client. proxy, codecs and pollers rewrite

### PR DESCRIPTION
**Pull request by hoa-grvt**
* [Refactor] `Archive` to use java state machine connection polling
  * `AsyncConnect` class `archive_async_connect.go` that does statemachine connection polling
  * `ArchiveContext` class - context for archive to connect, bridging from old aeron go init code to new async connect
  * Remove the homebrew `nextCorrelationID()` and use proper media driver `NextCorrelationID()` for id uniqueness.
* [Refactor/Rewrite] `Control`
  *  90-100% like Java implementation for all `PollForXXX` but nested under `Control` namespace
  * Clean up all unused and convoluted mismatch of 3 pollers results that combine into 1
  * 3 separate pollers and fields now dedicate to each poller type (`control_response_poller.go`, `recording_descriptor_poller.go`, `recording_subscription_descriptor_poller.go`)
  * Add a field to store archiveError that we can act on the next polling cycle (since most methods don't return error)
  * `AppendXXXDescriptorsConsumer` to match previous functionality when polling for descriptors
  * Rewrite the entire control poller using java implementation - single read instead without loop draining the queue.
  * Fix dispatching recording signal twice in archive control.
* [Refactor] `Proxy` 
  * Decouple from archive parents link (like Java)
  * Add addition param `controlSessionId`  to each method instead of getting from archive SessionID implicitly (like java)
  * `OfferOnce` is Java equivalent to Offer with no retry. Is used in function with `TryXXX` prefix
  * Add java like functions that requires boolean check on top of error (`KeepAlive`, `IsConnected`)
  * Add `proxy_compat.go` contains newer semantic version protocol that uses `archiveId` and `tryXXX` function
* [New] `codecs` - Sbe Version/Semantic Version Bump to `1.11.0` equivalent and support to `ArchiveId` request and response protocol

**Testing**

Async Connect State Machine + Local snapshot and tested with aeronheartbeat

```
------ New Archive Connect Polling (calling async connect underneath)
2024-09-25T23:22:29.111+0800    info    archive archive/archive.go:284  archive asyncConnect state transition: 0 -> 4
2024-09-25T23:22:29.116+0800    info    archive archive/archive.go:284  archive asyncConnect state transition: 4 -> 5
2024-09-25T23:22:29.116+0800    info    archive archive/archive.go:284  archive asyncConnect state transition: 5 -> 6
2024-09-25T23:22:29.117+0800    info    archive archive/archive.go:284  archive asyncConnect state transition: 6 -> 7



------ Async Connect Polling State Machine
2024-09-25T23:22:29.111+0800    debug   archive archive/archive_async_connect.go:109    Archive Connection State transition to: AWAIT_PUBLICATION_CONNECTED
2024-09-25T23:22:29.111+0800    debug   archive archive/archive_async_connect.go:117    Archive Connection State transition to: SEND_CONNECT_REQUEST
2024-09-25T23:22:29.111+0800    debug   archive archive/archive_async_connect.go:137    Archive Connection State transition to: AWAIT_SUBSCRIPTION_CONNECTED
2024-09-25T23:22:29.111+0800    debug   archive archive/archive_async_connect.go:146    Archive Connection State transition to: AWAIT_CONNECT_RESPONSE
2024-09-25T23:22:29.116+0800    debug   archive archive/archive_async_connect.go:234    Archive Connection State transition to: SEND_ARCHIVE_ID_REQUEST
2024-09-25T23:22:29.116+0800    debug   archive archive/archive_async_connect.go:159    Archive Connection State transition to: AWAIT_ARCHIVE_ID_RESPONSE
2024-09-25T23:22:29.117+0800    debug   archive archive/archive_async_connect.go:243    transitionToDone: archiveId=1 ControlSessionId=909925929
2024-09-25T23:22:29.117+0800    debug   archive archive/archive_async_connect.go:258    Archive Connection State transition to: DONE



------ Taking snapshot
2024-09-25T23:22:29.117+0800    debug   aeron   aeron/clientconductor.go:366    AddPublication: channel=aeron:ipc?alias=snapshot, streamId=1000
2024-09-25T23:22:29.228+0800    debug   aeron   aeron/clientconductor.go:848    Updated publication: &{87 87 1727277749117604438 1000 1852718933 133 -1 0 1 aeron:ipc?alias=snapshot  0xc000020000 <nil>}
2024-09-25T23:22:29.228+0800    info    archive archive/archive.go:131  NewPublicationListener(channel:aeron:ipc?alias=snapshot stream:1000, session:1852718933, regID:87)
2024-09-25T23:22:29.231+0800    debug   cluster cluster/clustered_service_agent.go:778  takeSnapshot - got recordingId: 1
2024-09-25T23:22:29.232+0800    info    aeronservice/snapshot.go:19     WritePartialSnapshot    {"type": "telemetry", "order_key": 0, "size": 56}
```

** References **
* `Poll` - https://github.com/real-logic/aeron/blob/master/aeron-archive/src/main/java/io/aeron/archive/client/ControlResponsePoller.java#L98
*  `onFragment` - https://github.com/real-logic/aeron/blob/1.46.2/aeron-archive/src/main/java/io/aeron/archive/client/ControlResponsePoller.java#L260
* `PollForResponse` -  https://github.com/real-logic/aeron/blob/1.46.2/aeron-archive/src/main/java/io/aeron/archive/client/AeronArchive.java#L2345
* `PollForErrorResponse` - https://github.com/real-logic/aeron/blob/1.46.2/aeron-archive/src/main/java/io/aeron/archive/client/AeronArchive.java#L440
* `pollNextResponse` - https://github.com/real-logic/aeron/blob/1.46.2/aeron-archive/src/main/java/io/aeron/archive/client/AeronArchive.java#L2309

### CODE REVIEW

File: archive/control.go [hunk 1]
```diff
@@ -46,11 +46,30 @@ type Control struct {
...
 // These pieces are filled out by various ResponsePollers which will set IsPollComplete to true
 type ControlResults struct {
-	CorrelationId                    int64
+	// ControlResponse
+	ControlSessionId int64
+	CorrelationId    int64
+	RelevantId       int64
+	Code             codecs.ControlResponseCodeEnum
+	Version          int32
+	ErrorMessage     []uint8
```
Comment (Aaronong) [inline archive/control.go]: Why not `string` or `[]byte`?
Java uses `String`

File: archive/control.go [hunk 2]
```diff
@@ -46,11 +46,30 @@ type Control struct {
...
+
+	// RecordingSignal
+	RecordingSignal codecs.RecordingSignalEnum
+	RecordingId     int64
+	SubscriptionId  int64
+	Position        int64
+
+	// Challenge
+	EncodedChallenge []uint8
```
Comment (Aaronong) [inline archive/control.go]: Java uses `byte[]` 

Why not `[]byte`?

File: archive/control.go [hunk 3]
```diff
@@ -46,11 +46,30 @@ type Control struct {
 // The polling mechanism is not parameterized so we need to set state for the results as we go
 // These pieces are filled out by various ResponsePollers which will set IsPollComplete to true
 type ControlResults struct {
-	CorrelationId                    int64
+	// ControlResponse
```
Comment (Aaronong) [inline archive/control.go]: I'm somewhat against this design. 

This overloads Java's `ControlResponsePoller`, `RecordingSignalPoller`, and `RecordingDescriptorPoller` into one class. 

I would prefer if these are separated into different structs.
For this PR, let's try to separate `ControlResponsePoller` out.

I see a couple hacks below that arises from sharing the class.

File: archive/control.go [hunk 4]
```diff
@@ -46,11 +46,30 @@ type Control struct {
...
+	SubscriptionId  int64
+	Position        int64
+
+	// Challenge
+	EncodedChallenge []uint8
+
+	// Poller complete status
+	TemplateId     int16
+	IsPollComplete bool
+
```
Comment (Aaronong) [inline archive/control.go]: The fields below are not in `ControlResponsePoller`

File: archive/control.go [hunk 5]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+	if control.Poll() != 0 && control.Results.IsPollComplete {
+		if control.Results.ControlSessionId == control.archive.SessionID {
+			if control.Results.Code == codecs.ControlResponseCode.ERROR {
+				archiveErr := NewArchiveError(control.Results.CorrelationId, int(control.Results.RelevantId), fmt.Sprintf("PollForErrorResponse received a ControlResponse (correlationId:%d Code:ERROR error=\"%s\")", control.Results.CorrelationId, control.Results.ErrorMessage))
+				control.Results.ErrorResponse = archiveErr
+				return aeron.NullValue, archiveErr
+			}
+
+			if control.Results.TemplateId == int16(codecIds.recordingSignalEvent) {
```
Comment (Aaronong) [inline archive/control.go]: else if

File: archive/control.go [hunk 6]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+
+// ----- Rewrite from Java ---- //
+
+// PollForErrorResponse polls the response stream once for an error
+// If another message is present then it will be skipped over
+// so only call when not expecting another response. If not connected then return NOT_CONNECTED_MSG
+// return the error otherwise nil if no error is found.
+func (control *Control) PollForErrorResponse() (int, error) {
+	control.Results.ErrorResponse = nil
+	control.Results.ErrorMessage = []uint8{}
```
Comment (Aaronong) [inline archive/control.go]: This is the issue with overloading the pollers I think.
we have random fields like this polluting. 

If we are sticking with this design, I think this should be done in `Control.Poll`

File: archive/control.go [hunk 7]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
 	return nil
 }
+
+// ----- Rewrite from Java ---- //
+
+// PollForErrorResponse polls the response stream once for an error
+// If another message is present then it will be skipped over
+// so only call when not expecting another response. If not connected then return NOT_CONNECTED_MSG
+// return the error otherwise nil if no error is found.
+func (control *Control) PollForErrorResponse() (int, error) {
```
Comment (Aaronong) [inline archive/control.go]: I think good to drop in comment linking to Java code.

This is `Archive.pollForErrorResponse`

File: archive/control.go [hunk 8]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+		}
+	}
+
+	// If we polled and did nothing then return
+	return aeron.NullValue, nil
+}
+
+// Poll for control response events. Returns number of fragments read during the operation.
+// Zero if no events are available.
+func (control *Control) Poll() (workCount int) {
```
Comment (Aaronong) [inline archive/control.go]: I think good to drop in comment linking to Java code.

This is `ControlResponsePoller.poll`

File: archive/control.go [hunk 9]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+			RecordingId:     aeron.NullValue,
+			SubscriptionId:  aeron.NullValue,
+			Position:        aeron.NullValue,
+
+			EncodedChallenge: []uint8{},
+
+			TemplateId:     aeron.NullValue,
+			IsPollComplete: false,
+
+			ControlResponse: nil,
```
Comment (Aaronong) [inline archive/control.go]: This is different

File: archive/control.go [hunk 10]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+			return fmt.Errorf("timeout waiting for correlationID %d", correlationID)
+		}
+
+		control.archive.Options.IdleStrategy.Idle(0)
+	}
+
+	return nil
+}
+
+func (control *Control) onFragment(
```
Comment (Aaronong) [inline archive/control.go]: I think good to drop in comment linking to Java code.

This is `ControlResponsePoller.onFragment`

File: archive/control.go [hunk 11]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+	marshaller := codecs.NewSbeGoMarshaller()
+	if err := hdr.Decode(marshaller, buf); err != nil {
+		// Not much to be done here as we can't correlate
+		err = fmt.Errorf("DescriptorFragmentHandler() failed to decode control message header: %w", err)
+		if control.archive.Listeners.ErrorListener != nil {
+			control.archive.Listeners.ErrorListener(err)
+		}
+		return term.ControlledPollActionContinue
+	}
+
```
Comment (Aaronong) [inline archive/control.go:291]: We are missing the `schemaID` check.

File: archive/control.go [hunk 12]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+			}
+
+			break
+		}
+
+		if fragments > 0 {
+			continue
+		}
+
+		if control.Subscription.IsClosed() {
```
Comment (Aaronong) [inline archive/control.go]: I don't think this is the same as `!control.Subscription.IsConnected()` 

I suggest we change.

File: archive/control.go [hunk 13]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+				logger.Debug(archiveErr) // log it in debug mode as an aid to diagnosis
+				return aeron.NullValue, archiveErr
+			}
+			return control.Results.RelevantId, nil
+		}
+	}
+}
+
+func (control *Control) pollNextResponse(startTime time.Time, correlationID, sessionID int64) error {
+	for {
```
Comment (Aaronong) [inline archive/control.go]: The idlestrategy reset. 

I think its as easy as `control.archive.Options.IdleStrategy.Idle(1)`

File: archive/control.go [hunk 14]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+		if fragments > 0 {
+			continue
+		}
+
+		if control.Subscription.IsClosed() {
+			return fmt.Errorf("response channel from archive is not connected")
+		}
+
+		// Check deadline
+		if time.Since(startTime) > control.archive.Options.Timeout {
```
Comment (Aaronong) [inline archive/control.go]: Why not follow the java implementation. 

Set deadline upfront. Then here we check `if time.After(deadline)`

File: archive/control.go [hunk 15]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+		if control.Results.Code == codecs.ControlResponseCode.ERROR {
+			archiveErr := NewArchiveError(correlationID, int(control.Results.RelevantId), fmt.Sprintf("response for correlationId=%d, error: %s", correlationID, control.Results.ErrorMessage))
+			logger.Debug(archiveErr) // log it in debug mode as an aid to diagnosis
+
+			if control.Results.CorrelationId == correlationID {
+				return aeron.NullValue, archiveErr
+			}
+		}
+
+		if control.Results.CorrelationId == correlationID {
```
Comment (Aaronong) [inline archive/control.go]: else if

File: archive/control.go [hunk 16]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+		}
+
+		if control.Results.ControlSessionId != control.archive.SessionID {
+			continue
+		}
+
+		// Check result
+		if control.Results.Code == codecs.ControlResponseCode.ERROR {
+			archiveErr := NewArchiveError(correlationID, int(control.Results.RelevantId), fmt.Sprintf("response for correlationId=%d, error: %s", correlationID, control.Results.ErrorMessage))
+			logger.Debug(archiveErr) // log it in debug mode as an aid to diagnosis
```
Comment (Aaronong) [inline archive/control.go]: I think these should be `logger.Error` not debug.

File: archive/control.go [hunk 17]
```diff
@@ -854,3 +718,252 @@ func (control *Control) PollForDescriptors(correlationID int64, sessionID int64,
...
+
+			if control.Results.CorrelationId == correlationID {
+				return aeron.NullValue, archiveErr
+			}
+		}
+
+		if control.Results.CorrelationId == correlationID {
+			if control.Results.Code != codecs.ControlResponseCode.OK {
+				archiveErr := NewArchiveError(correlationID, int(control.Results.RelevantId), "unexpected response code: "+string(control.Results.Code))
+				logger.Debug(archiveErr) // log it in debug mode as an aid to diagnosis
```
Comment (Aaronong) [inline archive/control.go]: `logger.Error`

File: archive/control.go [hunk 18]
```diff
@@ -418,6 +397,7 @@ func (control *Control) errorResponseFragmentHandler(buffer *atomic.Buffer, offs
 	switch hdr.TemplateId {
 	case codecIds.controlResponse:
 		var controlResponse = new(codecs.ControlResponse)
+		pollContext.control.Results.ControlResponse = controlResponse
```
Comment (Aaronong) [inline archive/control.go]: I think good to drop in comment linking to Java code.

Where is this from?

**Review by Aaronong — COMMENTED** [review]
This is a huge step up!
Some comments

File: archive/proxy_compat.go [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	ret := proxy.Offer(atomic.MakeBuffer(bytes, len(bytes)), 0, int32(len(bytes)), nil)
+	if ret < 0 {
+		return false, fmt.Errorf("Offer failed: %d", ret)
+	}
+
+	return ret > 0, nil
```
Comment (hoa-grvt) [inline archive/proxy_compat.go:22]: this one i am not 100% sure. 

Generally in the codebase proxy.Offer() is usually check >= 0 where 0 is a valid position. Java version though is checked against > 0

**Pull request merged**